httpx/middleware/auth.py

```diff
@@ -0,0 +1,219 @@
+import hashlib
+import os
+import time
+import typing
+from base64 import b64encode
+from urllib.request import parse_http_list
+
+from ..exceptions import ProtocolError
+from ..models import AsyncRequest, AsyncResponse, StatusCode
+from ..utils import to_bytes, to_str, unquote
+from .base import BaseMiddleware
+
+
+class BasicAuth(BaseMiddleware):
+    def __init__(
+        self, username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+    ):
+        if isinstance(username, str):
+            username = username.encode("utf-8")
+
+        if isinstance(password, str):
+            password = password.encode("utf-8")
```

Comment:
Thoughts on using to_bytes() here as well? :)

Reply:
Seems like that’d just be an obfuscation here.

Reply:
How is it different from what we’re doing in DigestAuth, though?

Reply:
Oh sorry, I missed that it was str_or_bytes. Yeah seems fair then.